### DISCUSSION TITLE
NextJS App Layout wont animate on exit but does animate on initial.

### DISCUSSION THREAD

**Discussion by ahmetskilinc** [motiondivision/motion#1614] [Q&A]
I currently have a NextJS app with a layout(`MainLayout.tsx`) that wraps my views. The layout has `motion.div` with `initial`, `animate` and `exit` on it with `layoutId="MainLayout"`.

The `AnimatePresence` is on the `_app.tsx` file with `exitBeforeEnter` on it. However it will not animate before exit, it only does the initial -> animate part.

```javascript
// _app.tsx
<AnimatePresence exitBeforeEnter>
  <Component {...pageProps} />
</AnimatePresence>
```

```javascript
// MainLayout.tsx
<motion.div
  layout
  initial={{ opacity: 0, y: -80 }}
  animate={{ opacity: 1, y: 0 }}
  exit={{ opacity: 0, y: 80 }}
  layoutId="MainLayout"
>
  <main>{props.children}</main>
</motion.div>
```

There is no other `motion` element/component in my app.

**Comment by harshagarwalnyu** [comment]
**Root cause:** `AnimatePresence` in `_app.tsx` wrapping `<Component>` works — but `MainLayout.tsx` is **inside** `Component`, so `AnimatePresence` sees the layout component unmount only when the route changes to a page with a *different* layout. If the layout persists across routes, `AnimatePresence` in `_app.tsx` never sees it unmount → no exit animation.

**Fix: move `AnimatePresence` inside `MainLayout.tsx` and key on route:**

```tsx
// _app.tsx — remove AnimatePresence here
export default function App({ Component, pageProps }) {
  return (
    <MainLayout>
      <Component {...pageProps} />
    </MainLayout>
  );
}
```

```tsx
// MainLayout.tsx
import { AnimatePresence, motion } from 'framer-motion';
import { useRouter } from 'next/router';

export default function MainLayout({ children }) {
  const router = useRouter();

  return (
    <div>
      {/* Static layout parts: nav, header, etc. */}
      <AnimatePresence mode="wait">
        <motion.div
          key={router.pathname}   // <-- THIS is what triggers exit
          initial={{ opacity: 0, y: -80 }}
          animate={{ opacity: 1, y: 0 }}
          exit={{ opacity: 0, y: 80 }}
          transition={{ duration: 0.3 }}
        >
          {children}
        </motion.div>
      </AnimatePresence>
    </div>
  );
}
```

**Why `key={router.pathname}` is critical:** Without it, React reuses the same DOM element across route changes (same `motion.div` never unmounts → `exit` never plays). With the key, React unmounts the old element and mounts a new one — `AnimatePresence` catches the unmount and plays `exit` before removal.

**Note:** `exitBeforeEnter` was renamed to `mode="wait"` in Framer Motion v7+. Use `mode="wait"` for v7+.

**Comment by carloitaben** [comment]
That's expected, learn more: https://github.com/vercel/next.js/discussions/42658
